src/Base/Tools.cpp

```diff
@@ -33,130 +33,217 @@
+    std::string nameSuffix;
+    uint digitCount;
+    uint digitsValue;
+    std::tie(digitCount, digitsValue) = decomposeName(name, baseName, nameSuffix);
+    baseName += nameSuffix;
+    auto baseNameEntry = UniqueSeeds.find(baseName);
+    if (baseNameEntry == UniqueSeeds.end()) {
+        // First use of baseName
+        baseNameEntry = UniqueSeeds
+                            .insert(std::pair<std::string, std::vector<PiecewiseSparseIntegerSet>>(
+                                baseName,
+                                std::vector<PiecewiseSparseIntegerSet>()))
+                            .first;
+    }
+    if (digitCount >= baseNameEntry->second.size()) {
+        // First use of this digitCount
+        baseNameEntry->second.resize(digitCount + 1);
+    }
+    PiecewiseSparseIntegerSet& baseNameAndDigitCountEntry = baseNameEntry->second[digitCount];
+    // Name should not already be there
+    assert(!baseNameAndDigitCountEntry.Contains(digitsValue));
+    baseNameAndDigitCountEntry.Add(digitsValue);
+}
+std::string Base::UniqueNameManager::makeUniqueName(const std::string& modelName,
+                                                    int minDigits) const
+{
+    std::string namePrefix;
+    std::string nameSuffix;
+    decomposeName(modelName, namePrefix, nameSuffix);
+    std::string baseName = namePrefix + nameSuffix;
+    auto baseNameEntry = UniqueSeeds.find(baseName);
+    if (baseNameEntry == UniqueSeeds.end()) {
+        // First use of baseName, just return it with no unique digits
+        return baseName;
+    }
+    // We don't care about the digit count of the suggested name, we always use at least the most
+    // digits ever used before.
+    int digitCount = baseNameEntry->second.size() - 1;
+    uint digitsValue;
+    if (digitCount < minDigits) {
+        // Caller is asking for more digits than we have in any registered name.
+        // We start the longer digit string at 000...0001 even though we might have shorter strings
+        // with larger numeric values.
+        digitCount = minDigits;
+        digitsValue = 1;
+    }
+    else {
+        digitsValue = baseNameEntry->second[digitCount].Next();
+    }
+    std::string digits = std::to_string(digitsValue);
+    if (digitCount > digits.size()) {
+        namePrefix += std::string(digitCount - digits.size(), '0');
+    }
+    return namePrefix + digits + nameSuffix;
 }
 
-std::string Base::Tools::addNumber(const std::string& name, unsigned int num, int d)
+// Remove a rgistered name so it can be generated again.
+// Nothing happens if you try to remove a non-registered name.
+void Base::UniqueNameManager::removeExactName(const std::string& name)
 {
-    std::stringstream str;
-    str << name;
-    if (d > 0) {
-        str.fill('0');
-        str.width(d);
+    std::string baseName;
+    std::string nameSuffix;
+    uint digitCount;
+    uint digitsValue;
+    std::tie(digitCount, digitsValue) = decomposeName(name, baseName, nameSuffix);
```

Comment:
got that from what little Python experience I have... I'm annoyed that it only works for `std::tuple` and not `std::pair`

Reply:
That's strange, it should work according to https://en.cppreference.com/w/cpp/utility/tuple/tie
> std::tie may be used to unpack a [std::pair](https://en.cppreference.com/w/cpp/utility/pair) because [std::tuple](https://en.cppreference.com/w/cpp/utility/tuple) has a [converting assignment](https://en.cppreference.com/w/cpp/utility/tuple/operator%3D) from pairs:
```c++
bool result;
std::tie(std::ignore, result) = set.insert(value);
```

```c++
std::set<S> set_of_s;
 
S value{42, "Test", 3.14};
std::set<S>::iterator iter;
bool is_inserted;
 
// Unpack a pair:
std::tie(iter, is_inserted) = set_of_s.insert(value);
assert(is_inserted);
```